process/process.go

```diff
@@ -741,20 +751,23 @@ func (p *Process) insertPropose(propose Propose) bool {
 		if p.broadcaster != nil {
 			p.broadcaster.BroadcastPrevote(Prevote{
 				Height: p.CurrentHeight,
-				Round:  p.CurrentRound,
+				Round:  propose.Round,
 				Value:  NilValue,
 				From:   p.whoami,
 			})
-			p.stepToPrevoting()
 		}
-		return false
+
+		p.ProposeLogs[propose.Round] = propose
+		p.ProposeIsValid[propose.Round] = false
+		return true
 	}
 
 	if _, ok := p.ProposeLogs[propose.Round]; ok {
 		return false
 	}
 
 	p.ProposeLogs[propose.Round] = propose
```

Comment:
This check needs to go immediately after the `propose.Round <= InvalidRound` check, otherwise a malicious proposer can force the node to do more work than necessary.

Reply:
Good point. I will also move the `CatchDoublePropose` with the `if _, ok := p.ProposeLogs[propose.Round]; ok` check.